nni/algorithms/compression/pytorch/pruning/transformer_pruner.py

```diff
@@ -0,0 +1,316 @@
+
+    Parameters
+    ----------
+    model : torch.nn.Module
+        Model to be pruned. Expect a model from transformers library (e.g., BertModel).
+        This pruner can work with other customized transformer models, but some ranking modes might fail.
+    config_list : list
+        Supported keys:
+            - sparsity : This is to specify the sparsity operations to be compressed to.
+            - op_types : Optional. Operation types to prune. (Should be 'Linear' for this pruner.)
+            - op_names : Optional. Operation names to prune.
+    head_hidden_dim : int
+        Dimension of the hidden dimension of each attention head. (e.g., 64 for BERT)
+        We assume that this head_hidden_dim is constant across the entire model.
+    attention_name_groups : list (Optional)
+        List of groups of names for weights of each attention layer. Each element should be a four-element list, with
+        the first three corresponding to Q_proj, K_proj, V_proj (in any order) and the last one being output_proj.
+    dummy_input : torch.Tensor (Optional)
+        Input to model's forward method, used to infer module grouping if attention_name_groups is not specified.
+        This tensor is used by the underlying torch.jit.trace to infer the module graph.
+    ranking_criterion : str
+        The criterion for ranking attention heads. Currently we support:
+            - l1_weight: l1 norm of Q_proj, K_proj, and V_proj
+            - l2_weight: l2 norm of Q_proj, K_proj, and V_proj
+            - l1_activation: l1 norm of the output of attention computation
+            - l2_activation: l2 norm of the output of attention computation
+            - taylorfo: l1 norm of the output of attention computation * gradient for this output
+                        (check more details in the masker documentation)
+    global_sort : bool
+        Whether rank the heads globally or locally before deciding heads to prune.
+    num_iterations : int
+        Number of pruning iterations. Defaults to 1 (ont-shot pruning). If num_iterations > 1, the pruner will split
+        the sparsity specified in config_list uniformly and assign a fraction to each pruning iteration.
+    epochs_per_iteration : int
+        Number of finetuning epochs before the next pruning iteration. This only has effect when num_iterations > 1.
+        If num_iterations is 1, then no finetuning is performed by the pruner after pruning.
+    optimizer: torch.optim.Optimizer
+        Optimizer used to train model
+    trainer: function
+        Function used to train the model.
+        Users should write this function as a normal function to train the Pytorch model
+        and include `model, optimizer, criterion, epoch` as function arguments.
+    criterion: function
+        Function used to calculate the loss between the target and the output.
+        For example, you can use ``torch.nn.CrossEntropyLoss()`` as input.
```

Comment:
Feel like that the `TransformerHeadPruner` is too heavy. I prefer to locate this pruner as a one-shot pruner, which means we do not need handle with the `num_iteration`, `optimizer`, 'criterion' things. That's much clearer. All those finetuning related things we can offload to the outer search algorithms. We can discuss with Quanlu @QuanluZhang .

Reply:
Yes, we can further discuss on that. One challenge is that this does not fit well in our current compression V1 framework (since the current iterative and dependency aware pruner are limited to convolutions), and compression V2 is not ready yet. My initial thought was to first integrate all these logic in one pruner (because of empirically good performance compared to one-shot pruning), and then factor out when compression V2 is ready.